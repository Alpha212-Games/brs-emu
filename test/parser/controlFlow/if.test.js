const Parser = require("../../../lib/parser");
const Expr = require("../../../lib/parser/Expression");
const { Lexeme } = require("../../../lib/Lexeme");
const BrsError = require("../../../lib/Error");
const { BrsBoolean, Int32 } = require("../../../lib/brsTypes");

const { token, identifier, EOF } = require("../ParserTests");

describe("parser if statements", () => {
    afterEach(() => BrsError.reset());

    describe("single-line if", () => {
        it("parses if only", () => {
            let parsed = Parser.parse([
                token(Lexeme.If),
                token(Lexeme.Integer, new Int32(1)),
                token(Lexeme.Less),
                token(Lexeme.Integer, new Int32(2)),
                token(Lexeme.Then),
                identifier("foo"),
                token(Lexeme.Equal),
                token(Lexeme.True, BrsBoolean.True),
                token(Lexeme.Newline),
                EOF
            ]);


            expect(BrsError.found()).toBe(false);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses if-else", () => {
            let parsed = Parser.parse([
                token(Lexeme.If),
                token(Lexeme.Integer, new Int32(1)),
                token(Lexeme.Less),
                token(Lexeme.Integer, new Int32(2)),
                token(Lexeme.Then),
                identifier("foo"),
                token(Lexeme.Equal),
                token(Lexeme.True, BrsBoolean.True),
                token(Lexeme.Else),
                identifier("foo"),
                token(Lexeme.Equal),
                token(Lexeme.True, BrsBoolean.False),
                token(Lexeme.Newline),
                EOF
            ]);

            expect(BrsError.found()).toBe(false);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it.skip("parses if-elseif-else", () => {
            // TODO: Figure out if
            // ```
            //     if a < b then foo = a else if a = b then foo = invalid else foo = b`
            // ```
            // is legal on a Roku device.  Specifically, is `else if` allowed in the single-line
            // form of if/then?
            let parsed = Parser.parse([
                token(Lexeme.If),
                token(Lexeme.Integer, new Int32(1)),
                token(Lexeme.Less),
                token(Lexeme.Integer, new Int32(2)),
                token(Lexeme.Then),
                identifier("foo"),
                token(Lexeme.Equal),
                token(Lexeme.True, BrsBoolean.True),
                token(Lexeme.ElseIf),
                token(Lexeme.Integer, new Int32(1)),
                token(Lexeme.Equal),
                token(Lexeme.Integer, new Int32(2)),
                identifier("same"),
                token(Lexeme.Equal),
                token(Lexeme.True, BrsBoolean.True),
                token(Lexeme.Else),
                identifier("foo"),
                token(Lexeme.Equal),
                token(Lexeme.True, BrsBoolean.False),
                token(Lexeme.Newline),
                EOF
            ]);

            expect(BrsError.found()).toBe(false);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});