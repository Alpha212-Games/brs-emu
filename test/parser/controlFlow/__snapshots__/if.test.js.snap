// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser if statements single-line if parses if only 1`] = `
Array [
  If {
    "condition": Binary {
      "left": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 1,
        },
      },
      "right": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 2,
        },
      },
      "token": Object {
        "kind": 15,
        "line": 1,
        "literal": undefined,
      },
    },
    "elseBranch": undefined,
    "elseIfs": Array [],
    "thenBranch": Expression {
      "expression": Binary {
        "left": Variable {
          "name": Object {
            "kind": 21,
            "line": 1,
            "text": "foo",
          },
        },
        "right": Literal {
          "value": BrsBoolean {
            "kind": 1,
            "value": true,
          },
        },
        "token": Object {
          "kind": 19,
          "line": 1,
          "literal": undefined,
        },
      },
    },
  },
]
`;

exports[`parser if statements single-line if parses if-else 1`] = `
Array [
  If {
    "condition": Binary {
      "left": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 1,
        },
      },
      "right": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 2,
        },
      },
      "token": Object {
        "kind": 15,
        "line": 1,
        "literal": undefined,
      },
    },
    "elseBranch": Expression {
      "expression": Binary {
        "left": Variable {
          "name": Object {
            "kind": 21,
            "line": 1,
            "text": "foo",
          },
        },
        "right": Literal {
          "value": BrsBoolean {
            "kind": 1,
            "value": true,
          },
        },
        "token": Object {
          "kind": 19,
          "line": 1,
          "literal": undefined,
        },
      },
    },
    "elseIfs": Array [],
    "thenBranch": Expression {
      "expression": Binary {
        "left": Variable {
          "name": Object {
            "kind": 21,
            "line": 1,
            "text": "foo",
          },
        },
        "right": Literal {
          "value": BrsBoolean {
            "kind": 1,
            "value": true,
          },
        },
        "token": Object {
          "kind": 19,
          "line": 1,
          "literal": undefined,
        },
      },
    },
  },
]
`;

exports[`parser if statements single-line if parses if-elseif-else 1`] = `Array []`;
